Anchoring: added tests for two-phase anchoring

csillag opened the pull request:
Develop the test suite for the anchoring system:

add tests for two-phase a.k.a. lazy anchoring,
covering the on-the-fly creation and removal of highlights,
events announcing the creation/removal of highlights,
testing the realized/virtual state of anchors, etc.

(A prelude for testing the two-phase scrolling feature.)

csillag commented:
Hey, my build seems to be failing because of a [seemingly independent problem](https://travis-ci.org/hypothesis/h/builds/50131564#L971). Does any of you have any idea about this? I have not touched server side at all in this PR... but according to the history, the relevant file is 2 months old. Yet, it has never triggered a warning until now. 

...which clearly means that I am misinterpreting the Travis output. Could someone please give check my sleepy eyes?

Thanks!

aron commented:
@csillag these errors are usually triggered by lint errors, this happens occasionally when one of the tools used by prospector is updated to include new rules. In this case it's now picking up an inconsistent whitespace error in server.py.

https://travis-ci.org/hypothesis/h/builds/50131564#L965-L988

Best thing to do in these situations is to find the error, fix it on master then rebase you current branch ontop of master. This should now be fixed by 55d8463.

aron commented on tests/js/plugin/enhancedanchoring-test.coffee:
```diff
@@ -106,6 +110,17 @@ describe 'Annotator.Plugin.EnhancedAnchoring', ->
+      anchor = am.createAnchor(ann, ann.target[0]).result
+
+      hl = anchor.highlight[anchor.startPage]
+      assert am.annotator.publish.calledWith "highlightsCreated", [hl]
```
Sinon provides it's own [assertions](http://sinonjs.org/docs/#assertions) for spies that give more detailed output in the error messages. We usually mix them into the chai `assert` object at the top of the test file.

```coffee
assert = chai.assert
sinon.assert.expose(assert, prefix: '')
```

They can then be used anywhere in the test file, for example:

```coffee
assert.calledWith(am.annotator.publish, 'highlightsCreated', [hl])
```

csillag replied:
On 2015-02-10 10:23, Aron Carroll wrote:
> Sinon provides it's own assertions
> <http://sinonjs.org/docs/#assertions> for spies that give more
> detailed output in the error messages.

OK, I migrated to using those.

> We usually mix them into the chai |assert| object at the top of the
> test file.

That mixing is magic.

aron commented on tests/js/plugin/enhancedanchoring-test.coffee:
```diff
@@ -202,3 +217,224 @@ describe 'Annotator.Plugin.EnhancedAnchoring', ->
+      assert.isObject anchor.highlight
+      assert.notOk anchor.highlight[1], "Should not have a highlight on page 1"
+
+      assert am.annotator.publish.notCalled
```
Generally each test case should assert just one thing. In this case I think we would benefit from splitting this up into three separate cases.

1) Assert anchor isn't realised.
2) Assert that no pages are rendered.
3) Assert that no event is triggered.

csillag replied:
On 2015-02-10 10:26, Aron Carroll wrote:
>
> Generally each test case should assert just one thing. In this case I
> think we would benefit from splitting this up into three separate cases.
>
> 1) Assert anchor isn't realised.
> 2) Assert that no pages are rendered.
> 3) Assert that no event is triggered.
>

OK, I am willing to split, but that would result in many useless code
repetition. (The steps of the scenario need to be written again and again.)
Should I not be bothered by that? Or I could put those steps into helper
functions, but earlier you said that you would prefer to see
in each steps what's really going on.

So, should I simply be repeating the same steps again and again, in
order to check different aspects of the results?

aron replied:
Repetition in tests is fine if it improves readability for others coming to the tests later on and clarity of errors when running the tests.

My general rule of thumb is to repeat setup in the method if its just a few lines, otherwise look at ways of moving it into a setup helper. The more contained the test the better, so it's a matter of balance.

aron commented on tests/js/plugin/enhancedanchoring-test.coffee:
```diff
@@ -202,3 +217,224 @@ describe 'Annotator.Plugin.EnhancedAnchoring', ->
+        event = document.createEvent "UIEvents"
+        event.initUIEvent "docPageUnmapped", false, false, window, 0
+        event.pageIndex = index
+        window.dispatchEvent event
```
These helper functions are not part of the `DummyDocumentAccess` object so it would be clearer to just move them into helpers that accept a `document`. e.g.

```coffee
unrenderPage: (doc, index) ->
  return unless doc.isPageMapped(index)
  ...
```

csillag replied:
On 2015-02-10 10:46, Aron Carroll wrote:
> These helper functions are not part of the |DummyDocumentAccess| object 

In fact they kind of are, in the sense that normally, it's the
responsibility of the document access policy to monitor the platform
(which is doing the actual rendering), and to notify Annotator (or, to
be more concrete, the anchoring subsystem) about any page rendering events.

So these events are generally emitted by the document access policy;
that's why I put it there.

(But technically, I could just as well as put them into a separate
function, if you would prefer that.)

aron replied:
Okay, that's true, but in this case it makes it harder to distinguish which properties on the object are mocking the implementation and which are part of the test suite.

A mock object will ideally be the simplest thing that acts as a stand-in for the real object. Ideally it will just be a plain object that has the correct state, or stubs that return the correct values for a particular test. The less logic the better.

csillag replied:
On 2015-02-10 13:10, Aron Carroll wrote:
>
> Okay, that's true, but in this case it makes it harder to distinguish
> which properties on the object are mocking the implementation and
> which are part of the test suite.
>
> A mock object will ideally be the simplest thing that acts as a
> stand-in for the real object. Ideally it will just be a plain object
> that has the correct state, or stubs that return the correct values
> for a particular test. The less logic the better.
>

OK, I have separated the helper functions.

aron commented:
Looking good, as I commented above, I think a number of the tests could be split up into smaller groups of assertions to help with readability. It also helps with debugging e.g. when a test fails you don't need to work out which of the 7+ assertions is failing.

csillag commented:
On 2015-02-10 10:13, Aron Carroll wrote:
>
> @csillag <https://github.com/csillag> these errors are usually
> triggered by lint errors, this happens occasionally when one of the
> tools used by prospector is updated to include new rules. In this case
> it's now picking up an inconsistent whitespace error in server.py.
>
> https://travis-ci.org/hypothesis/h/builds/50131564#L965-L988
>

Yeah, I saw that - I pointed to the same error in my original travis
link. (btw, travis fails to scrolls to the line when you open the link,
because it's not yet rendered when the browser window opens. They need
two-phase scrolling :)

> Best thing to do in these situations is to find the error,
>

I tried, but
 - prospector doesn't seem to be installed by default.
 - even after I installed it manually, it gave me a completely different
output...

> fix it on master then rebase you current branch ontop of master. This
> should now be fixed by 55d8463
> <https://github.com/hypothesis/h/commit/55d84631022f2986d40c021d4cd6fe06555b7274>.
>

Thanks! After rebasing, the error is indeed gone. It's all green.

aron commented:
Okay, I added documentation to the README about installing/running the linter. https://github.com/hypothesis/h/blob/master/README.rst#code-quality

csillag commented:
Thanks!

csillag commented:
On 2015-02-10 10:50, Aron Carroll wrote:
>
> Looking good, as I commented above, I think a number of the tests
> could be split up into smaller groups of assertions to help with
> readability. It also helps with debugging e.g. when a test fails you
> don't need to work out which of the 7+ assertions is failing.
>

OK, I separated most of the test cases in to teeny-weeny bits.
(And I believe that I have addressed all the comments so far.)

aron commented on tests/js/plugin/enhancedanchoring-test.coffee:
```diff
@@ -202,3 +218,404 @@ describe 'Annotator.Plugin.EnhancedAnchoring', ->
+
+    # Helper function to trigger a page unrendering
+    unrenderPage = (doc, index) ->
+      return unless doc.isPageMapped index
```
Let's get rid of these statements, they're unnecessary if we set up the test correctly.

csillag replied:
> Let's get rid of these statements,

You mean the checks against calling these helper functions in the incorrect state?

> they're unnecessary if we set up the test correctly.

Yes, they are not necessary in the strict sense, but if remove them, and someone write incorrect tests, he can trigger weird things that never happen normally. (Like, rendering of a page that is already rendered.) Shouldn't we have some protection against that?

aron replied:
At least that way you get inconsistent behaviour instead of nothing happening. But a better way of dealing with bad input is just to throw an error if someone passes the wrong index.

csillag replied:
On 2015-02-10 15:27, Aron Carroll wrote:
>
> At least that way you get inconsistent behaviour instead of nothing
> happening. But a better way of dealing with bad input is just to throw
> an error if someone passes the wrong index.
>

Yes, that's indeed better.

I added exceptions instead.

aron commented:
Looks great! One last comment about the helpers, then we're good to go.

csillag commented:
> Looks great! One last comment about the helpers, then we're good to go.

Tell me when you feel that this is ready; I want to squash it before merging.

csillag commented:
> Tell me when you feel that this is ready; I want to squash it before merging.

In fact I squashed it right now.

aron commented on tests/js/plugin/enhancedanchoring-test.coffee:
```diff
@@ -202,3 +218,407 @@ describe 'Annotator.Plugin.EnhancedAnchoring', ->
+    # Helper function to trigger a page rendering
+    renderPage = (doc, index) ->
+      if doc.isPageMapped(index)
+        throw new Error "That page is already rendered!"
```
Perhaps something more descriptive here so you actually know what went wrong?

    "Cannot call renderPage with an already mapped index: ${index}, ensure the document is setup correctly"
    "Cannot call unrenderPage with an unmapped index: ${index}, ensure the document is setup correctly"

csillag replied:
On 2015-02-10 15:38, Aron Carroll wrote:
>
> Perhaps something more descriptive here so you actually know what went
> wrong?
>
Done.

csillag commented:
Squashed again, in the (vain) hope that this was the last change.

aron commented:
Nice!

tilgovi commented:
@csillag nice
